core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthDocRequest.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2018-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.stitch.core.internal.net;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.mongodb.stitch.core.StitchRequestErrorCode;
+import com.mongodb.stitch.core.StitchRequestException;
+import com.mongodb.stitch.core.internal.common.StitchObjectMapper;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.bson.Document;
+
+public final class StitchAuthDocRequest extends StitchAuthRequest {
+  private final Document document;
+
+  public StitchAuthDocRequest(final StitchAuthRequest request, final Document document) {
+    super(request);
+    this.document = document;
+  }
+
+  public StitchAuthDocRequest(final StitchRequest request, final Document document) {
+    super(request, false);
+    this.document = document;
+  }
+
+  public Builder builder() {
+    return new Builder(this);
+  }
+
+  public Document getDocument() {
+    return document;
+  }
+
+  public static class Builder extends StitchAuthRequest.Builder {
+    private Document document;
+
+    public Builder() {}
+
+    Builder(final StitchAuthDocRequest request) {
+      super(request);
+      document = request.document;
+    }
+
+    public Builder withDocument(final Document document) {
+      this.document = document;
+      return this;
+    }
+
+    public Document getDocument() {
+      return this.document;
+    }
+
+    /**
+     * Set if this request should use an access token in this request.
+     */
+    public Builder withAccessToken() {
```

Comment:
You'll see this pattern of overriding and using super (without the annotation) to indicate that this is purely for allowing you to do `builder.withSuperMethod().withConcreteMethod().withSuperMethod().build()` and still get back the concrete built type. CRTP (https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) didn't work out here and it confuses me very much in general :)